### DISCUSSION TITLE
Custom links with MAC address

### DISCUSSION THREAD

**Discussion by Azmodeszer** [netbox-community/netbox#14553] [Q&A] [answered]
I'm trying to set up a custom link that incorporates a device's MAC address in order to link to an external monitoring GUI. I have this so far:

```
{% set mac_address = '' %}
{% if object.interfaces.get(name='GigabitEthernet0/2') %}
    {% set mac_address = object.interfaces.get(name='GigabitEthernet0/2').mac_address %}
{% elif object.interfaces.get(name='GigabitEthernet0') %}
    {% set mac_address = object.interfaces.get(name='GigabitEthernet0').mac_address %}
{% endif %}

{% if mac_address %}
    https://<webguiIP>/screens/dashboard.html#!/APTable/APView/{{ mac_address }}
{% else %}
    No MAC address found for specified interfaces.
{% endif %}
```

This *almost* works, but there are two hurdles:

1) It doesn't seem to parse the elif condition, i.e. unless the interface is named "GigabitEthernet0/2", it says that no interface was found, even if there is an interface named "GigabitEthernet0". Either way this isn't very elegant. Ideally, it should just iterate through all the attached interfaces and select the MAC address where the value is not NULL or similar (the relevant interface should always be the one with a MAC address entered). I'm not a coder and mostly got the code above from ChatGPT and some basic logic/trial&error, so any guidance appreciated. The official documentation does not go into enough detail to help me with this, it seems.

2) The ! in the URL is always encoded as "%21" in the resulting link, which renders it useless. Is there a way to enforce the "!"?

**Comment by candlerb** [comment, marked as answer]
> The ! in the URL is always encoded as "%21" in the resulting link, which renders it useless. Is there a way to enforce the "!"?

What's your Netbox version? I remember seeing there was a change in URL encoding recently. I think it's #13011. Maybe you need to make a feature request to allow this character as well. (FWIW, exclamation mark is a "sub-delim" in rfc3986. I think reserved character escaping should be done inside `{{ .. }}` expansions, but not in the static part of a URL)

Regarding database queries: the interface you are using is the Django ORM, so you can look at the Django documentation for this.

For finding the first interface with a MAC address you probably want something like `object.interfaces.filter(mac_address__isnull=False).first()` (although if there are multiple interfaces like this it may be indeterminate which one is picked)

The reason your `elif` clause isn't being used is probably because the first `get()` which doesn't find any object raises an exception, rather than returning None, so your whole template execution will abort.

The best way to test all this interactively is using [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/).

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
### NetBox interactive shell (netbox3)
### Python 3.8.10 | Django 4.2.8 | NetBox 3.6.7
### lsmodels() will show available models. Use help(<model>) for more info.
>>> o = Device.objects.get(name="nuc1")
>>> o.interfaces.filter(mac_address__isnull=False).first()
<Interface: br255>
>>> o.interfaces.filter(mac_address__isnull=False).first().mac_address
EUI('AA:BB:CC:DD:EE:FF')
>>>
```

but:

```
>>> o.interfaces.get(name="nonexistent")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/models/query.py", line 637, in get
    raise self.model.DoesNotExist(
dcim.models.device_components.Interface.DoesNotExist: Interface matching query does not exist.
```

Note that unlike `get()`, `first()` returns None if there is no match, which makes it easier to use.

```
>>> o.interfaces.filter(mac_address="ff:ff:ff:ff:ff:ff").first()
>>>
```

**Reply by Azmodeszer** [reply]
I'm using v3.6.6.

Thanks for the response, I'll have a look at your suggestions.

**Reply by Azmodeszer** [reply]
Cheers, seems to work fine for my purposes with

`https://<URL>/{{ object.interfaces.filter(mac_address__isnull=False).first().mac_address }}`

Apart from the ! thing, of course. Should I open an issue/feature request?

**Reply by candlerb** [reply]
Yes I think a feature request is the right thing here.

**Reply by Azmodeszer** [reply]
Done.

One more thing: Is there a way to add a condition so that the Custom Link only shows up for devices with a specific Device Role?

**Reply by candlerb** [reply]
https://docs.netbox.dev/en/stable/customization/custom-links/#conditional-rendering

(Note: that means you need to use Jinja2 template for the link *text* as well as the link *URL*)
